xla/service/gpu/runtime/topk_kernel.cu.h

```diff
@@ -254,25 +217,33 @@ class TopK {
 // instantiations of Run() from the multiple monomorphizations of Run().
 extern __device__ __shared__ int shmem[];
 
-template <size_t K, typename KT, typename VT>
+template <size_t K, typename KT>
 __launch_bounds__(kTopKMaxThreadsPerBlock, 1) __global__
-    void Run(KT* data, int n, KT* result, uint32_t* result_idxs, int k) {
-  TopK<K, KT, VT> top_k(shmem, k);
+    void Run(KT* data, int n, KT* result, uint32_t* result_idxs, int k) 
+{
+  TopK<K, KT> obj(shmem, k);
+  
+  const uint32_t bidx = blockIdx.x; 
+  auto in = data + n * bidx;
+  auto vals_out = result + k * bidx;
+  auto idxs_out = result_idxs + k * bidx;
   int slice_size = n / blockDim.x;
   if (threadIdx.x < n % blockDim.x) {
     slice_size++;
   }
-  top_k.Run(&data[n * blockIdx.x], slice_size, &result[k * blockIdx.x],
-            &result_idxs[k * blockIdx.x]);
+ 
+  obj.PerWarpTopK(in, slice_size);
+  obj.MergeTopKs(vals_out, idxs_out);
 }
 
 template <typename T, size_t K>
 void* GetTopKKernelForK(int n) {
   // TODO(doak): Switch to uint32_t if we don't have an efficient
   // implemementation for uint16_t.
-  return n < std::numeric_limits<uint16_t>::max()
-             ? reinterpret_cast<void*>(&Run<K, T, uint16_t>)
-             : reinterpret_cast<void*>(&Run<K, T, uint32_t>);
+  // pemeliya: gpu registers are 32-bits hence there is no performance
```

Comment:
I got a review comment from the original author of the code, he said
"There was a significant performance benefit when both key & value are 16-bit values."
So we should probably leave this as it was before.

Reply:
To be clear, what was being asked was: Leave the Traits typename for TopK and the Descending struct.
I think this will result in quite some rework of the code, and more than I would be willing to adjust myself.

Reply:
OK, I see, fair enough.. That I will have to return back the old data structures

Reply:
I have rewound back my changes regarding Descending Traits class. 
@akuegel : I am also wondering do you know a proper way to run google benchmark tests integrated into a google ?
For that purpose, I had to integrate the macro 
```
#define RUN_BENCHMARKS_ONLY 0
```
into topk_kernel_test.cc since otherwise google test framework inserts it's own 'main()' function which prohibits running the benchmarks. I guess, though, there should be a better way to do this ?

Reply:
Regarding the benchmarks: I think it should work that instead of bazel test ... you use
```bazel run  -- --benchmark_filter=all``` (i.e. pass the flag benchmark_filter to the binary).

Any yes, this macro also got a comment in the internal review. I guess if the bazel run command works for you, then this will not be needed?

Reply:
Indeed, this flag did the job ! thanks for the tip, I have removed the macro